prog/dftb+/prg_dftb/main.F90

```diff
@@ -0,0 +1,5637 @@
+            call writeDetailedOut3(fdUser, tPrintForces, tSetFillingTemp, tPeriodic, tStress,&
+                & totalStress, totalLatDeriv, energy, tempElec, pressure, cellPressure, kT)
+          end if
+        else if (tSocket .and. iGeoStep < nGeoSteps) then
+          ! Only receive geometry from socket, if there are still geometry iterations left
+          call receiveGeometryFromSocket(socket, tPeriodic, coord0, latVec, tCoordsChanged,&
+              & tLatticeChanged, tStopDriver)
+        end if
+      end if
+
+      if (tWriteDetailedOut) then
+        call writeDetailedOut4(fdUser, tMD, energy, kT)
+      end if
+
+      tStopDriver = tStopScc .or. tStopDriver .or. hasStopFile(fStopDriver)
+      if (tStopDriver) then
+        exit lpGeomOpt
+      end if
+
+    end do lpGeomOpt
+
+    if (tSocket) then
+      call socket%shutdown()
+    end if
+
+    tGeomEnd = tMD .or. tGeomEnd .or. tDerivs
+
+    if (tWriteDetailedOut) then
+      call writeDetailedOut5(fdUser, tGeoOpt, tGeomEnd, tMd, tDerivs, tEField, absEField,&
+          & dipoleMoment)
+    end if
+
+    call writeFinalDriverStatus(tGeoOpt, tGeomEnd, tMd, tDerivs)
+
+    if (tMD) then
+      call writeMdOut3(fdMd, mdOut)
+    end if
+
+    if (tDerivs) then
+      call getHessianMatrix(derivDriver, pDynMatrix)
+      call writeHessianOut(fdHessian, hessianOut, pDynMatrix)
+    else
+      nullify(pDynMatrix)
+    end if
+
+
+    if (allocated(pipekMezey)) then
+      ! NOTE: the canonical DFTB ground state orbitals are over-written after this point
+      if (tStoreEigvecs) then
+        call error("Pipek-Mezey localisation not implemented for stored eigenvectors")
+      end if
+      if (nSpin > 2) then
+        call error("Pipek-Mezey localisation not implemented for non-colinear DFTB")
+      end if
+      call calcPipekMezeyLocalisation(pipekMezey, nEl, filling, over, kPoint, kWeight,&
+          & neighborList, nNeighbor, iAtomStart, iPair, img2CentCell, iCellVec, cellVec, fdEigvec,&
+          & runId, orb, species, speciesName, localisation, HSqrReal, SSqrReal, HsqrCplx, SSqrCplx)
+    end if
+
+    if (tWriteAutotest) then
+      if (tPeriodic) then
+        cellVol = abs(determinant33(latVec))
+        energy%EGibbs = energy%EMermin + pressure * cellVol
+      end if
+      call writeAutotestTag(fdAutotest, autotestTag, tPeriodic, cellVol, tMulliken, qOutput,&
+          & totalDeriv, chrgForces, excitedDerivs, tStress, totalStress, pDynMatrix,&
+          & energy%EMermin, pressure, energy%EGibbs, coord0, tLocalise, localisation)
+    end if
+    if (tWriteResultsTag) then
+      call writeResultsTag(fdResultsTag, resultsTag, totalDeriv, chrgForces, tStress, totalStress,&
+          & pDynMatrix, tPeriodic, cellVol)
+    end if
+    if (tWriteDetailedXML) then
+      call writeDetailedXml(runId, speciesName, species0, pCoord0Out, tPeriodic, latVec, tRealHS,&
+          & nKPoint, nSpin, size(eigen, dim=1), nOrb, kPoint, kWeight, filling, occNatural)
+    end if
```

Comment:
We could/should move this two writing routines into the geometry loop indeed. But as in this branch I did not want to change program behaviour, let's do it later in an other branch.